vql/tools/sftp_upload.go

```diff
@@ -0,0 +1,202 @@
+	"golang.org/x/net/context"
+	"www.velocidex.com/golang/velociraptor/file_store/api"
+	"www.velocidex.com/golang/velociraptor/glob"
+	vql_subsystem "www.velocidex.com/golang/velociraptor/vql"
+	"www.velocidex.com/golang/vfilter"
+)
+
+type SFTPUploadArgs struct {
+	File       string `vfilter:"required,field=file,doc=The file to upload"`
+	Name       string `vfilter:"optional,field=name,doc=The name of the file that should be stored on the server"`
+	User       string `vfilter:"required,field=user,doc=The username to connect to the endpoint with"`
+	Path       string `vfilter:"required,field=path,doc=Path on server to upload file to"`
+	Accessor   string `vfilter:"optional,field=accessor,doc=The accessor to use"`
+	PrivateKey string `vfilter:"required,field=privatekey,doc=The private key to use"`
+	Endpoint   string `vfilter:"required,field=endpoint,doc=The Endpoint to use"`
+	HostKey    string `vfilter:"optional,field=hostkey,doc=Host key to verify. Blank to disable"`
+}
+
+type SFTPUploadFunction struct{}
+
+func (self *SFTPUploadFunction) Call(ctx context.Context,
+	scope *vfilter.Scope,
+	args *ordereddict.Dict) vfilter.Any {
+
+	arg := &SFTPUploadArgs{}
+	err := vfilter.ExtractArgs(scope, args, arg)
+	if err != nil {
+		scope.Log("upload_sftp: %s", err.Error())
+		return vfilter.Null{}
+	}
+
+	err = vql_subsystem.CheckFilesystemAccess(scope, arg.Accessor)
+	if err != nil {
+		scope.Log("upload_SFTP: %s", err)
+		return vfilter.Null{}
+	}
+
+	accessor, err := glob.GetAccessor(arg.Accessor, scope)
+	if err != nil {
+		scope.Log("upload_SFTP: %v", err)
+		return vfilter.Null{}
+	}
+
+	file, err := accessor.Open(arg.File)
+	if err != nil {
+		scope.Log("upload_SFTP: Unable to open %s: %s",
+			arg.File, err.Error())
+		return &vfilter.Null{}
+	}
+	defer file.Close()
+
+	if arg.Name == "" {
+		arg.Name = arg.File
+	}
+
+	stat, err := file.Stat()
+	if err != nil {
+		scope.Log("upload_SFTP: Unable to stat %s: %v",
+			arg.File, err)
+	} else if !stat.IsDir() {
+		// Abort uploading when the scope is destroyed.
+		sub_ctx, cancel := context.WithCancel(ctx)
+		scope.AddDestructor(func() {
+			cancel()
+		})
+
+		upload_response, err := upload_SFTP(
+			sub_ctx, scope, file,
+			strings.TrimSpace(arg.User),
+			strings.TrimSpace(arg.Path),
+			strings.TrimSpace(arg.Name),
+			strings.TrimSpace(arg.PrivateKey),
+			strings.TrimSpace(arg.Endpoint),
+			strings.TrimSpace(arg.HostKey))
+		if err != nil {
+			scope.Log("upload_SFTP: %v", err)
+			// Relay the error in the UploadResponse
+			return upload_response
+		}
+		return upload_response
+	}
+
+	return vfilter.Null{}
+}
+
+func keyString(k ssh.PublicKey) string {
+	return k.Type() + " " + base64.StdEncoding.EncodeToString(k.Marshal())
+}
+
+func hostkeycallback(trustedkey string) ssh.HostKeyCallback {
+	return func(_ string, _ net.Addr, k ssh.PublicKey) error {
+		ks := keyString(k)
+		if trustedkey != ks {
+			return fmt.Errorf("SSH-key verification: expected %q but got %q", trustedkey, ks)
```

Comment:
I think this %q will just print the address of the pointer so it is not very useful for debugging when things go wrong. maybe reuse keyString here? Or sometimes simply %v will use a type provided nice formatting of the type so ssh.PublicKey might already have one